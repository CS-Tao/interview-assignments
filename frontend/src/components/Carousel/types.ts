export interface CarouselItem {
  title: string | string[];
  subTitle?: string | string[];
  image: string;
  backColor: string;
  foreColor: string;
}
