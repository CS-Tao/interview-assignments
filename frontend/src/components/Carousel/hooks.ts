import { useEffect, useRef, useState } from "react";
import { CarouselItem } from "./types";

export const useSlide = (
  items: CarouselItem[],
  interval: number,
  animationDuration: number,
  onChange?: (item: CarouselItem, index: number) => void
): [number, boolean] => {
  const [index, setIndex] = useState(0);
  const [isFirstRender, seIsFirstRender] = useState(true);
  const onChangeRef = useRef(onChange);
  const timer = useRef<NodeJS.Timeout | null>();
  const nextIndex = (index + 1) % items.length;

  useEffect(() => {
    seIsFirstRender(false);
    timer.current = setTimeout(() => {
      setIndex(nextIndex);
      onChangeRef.current?.(items[nextIndex], nextIndex);
    }, interval + animationDuration);

    return () => {
      if (timer.current) {
        clearTimeout(timer.current);
        timer.current = null;
      }
    };
  }, [items, interval, animationDuration, nextIndex]);

  return [index, isFirstRender];
};
