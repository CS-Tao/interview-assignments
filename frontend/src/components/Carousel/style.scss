.carousel-wrapper {
  position: relative;
  overflow: hidden;

  .carousel-item {
    position: absolute;
    display: block;
    width: 100%;
    height: 100%;
    overflow: hidden;

    & > .title-wrapper {
      position: relative;
      z-index: 1;
      margin-top: 20vh;
      & > div {
        font-size: 56px;
        font-weight: bold;
        white-space: pre;
      }
    }

    & > .subtitle-wrapper {
      position: relative;
      z-index: 1;
      margin-top: 10px;
      & > div {
        font-size: 18px;
        white-space: pre;
      }
    }

    & > .image-wrapper {
      position: absolute;
      left: 0;
      top: 0;
      width: 100%;
      height: 100%;
      & > img {
        position: absolute;
        height: 50%;
        bottom: 140px;
        left: 50%;
        transform: translateX(-50%);
      }
    }
  }

  & > .carousel-indicator {
    position: absolute;
    bottom: 100px;
    left: 50%;
    transform: translateX(-50%);
    display: flex;
    justify-content: center;
    align-items: center;

    & > .carousel-indicator-item {
      width: 50px;
      height: 4px;
      border-radius: 5px;
      background-color: #a9a9a9;
      margin: 0 6px;

      &.active {
        background-color: #fbfafa;
      }

      & > .carousel-indicator-item-handler {
        height: 100%;
        border-radius: 5px;
        background-color: #fbfafa;
      }
    }
  }
}
