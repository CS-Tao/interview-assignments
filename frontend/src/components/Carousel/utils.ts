export const formatList = (item: string | string[]): string[] => {
  if (Array.isArray(item)) {
    return item;
  }
  return [item];
};
