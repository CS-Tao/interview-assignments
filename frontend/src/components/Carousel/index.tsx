import React, { CSSProperties, FC } from "react";
import { CarouselItem } from "./types";
import { formatList } from "./utils";
import { useSlide } from "./hooks";
import "./style.scss";

export interface CarouselProps {
  items: CarouselItem[];
  width?: CSSProperties["width"];
  height?: CSSProperties["height"];
  interval?: number;
  animationDuration?: number;
  onChange?: (item: CarouselItem, index: number) => void;
}

const Carousel: FC<CarouselProps> = (props) => {
  const {
    items,
    width = "100%",
    height = "100%",
    interval = 2000,
    animationDuration = 500,
    onChange,
  } = props;

  const [currentIndex, isFirstRender] = useSlide(
    items,
    interval,
    animationDuration,
    onChange
  );

  const renderItem = (item: CarouselItem, index: number) => {
    const { title, subTitle = [], image } = item;
    let left = `${(index - currentIndex) * 100}%`;

    return (
      <div
        className="carousel-item"
        style={{
          color: item.foreColor,
          backgroundColor: item.backColor,
          transition: `left ${animationDuration}ms linear`,
          left,
        }}
        key={index}
      >
        {renderTitle(formatList(title))}
        {renderSubTitle(formatList(subTitle))}
        {renderImage(image)}
      </div>
    );
  };

  const renderTitle = (titleList: string[]) => {
    return (
      <div className="title-wrapper">
        {titleList.map((title, index) => {
          return <div key={index}>{title}</div>;
        })}
      </div>
    );
  };

  const renderSubTitle = (subTitleList: string[]) => {
    return (
      <div className="subtitle-wrapper">
        {subTitleList.map((title, index) => {
          return <div key={index}>{title}</div>;
        })}
      </div>
    );
  };

  const renderImage = (image: string) => {
    return (
      <div className="image-wrapper">
        <img src={image} alt="" />
      </div>
    );
  };

  const renderIndicator = () => {
    return (
      <div className="carousel-indicator">
        {items.map((_, index) => {
          const active = index === currentIndex;

          return (
            <div key={index} className="carousel-indicator-item">
              <div
                className="carousel-indicator-item-handler"
                style={{
                  width: isFirstRender || !active ? "0%" : "100%",
                  transition: active
                    ? `width ${interval + animationDuration}ms linear`
                    : "",
                }}
              />
            </div>
          );
        })}
      </div>
    );
  };

  return (
    <div className="carousel-wrapper" style={{ width, height }}>
      {items.map(renderItem)}
      {renderIndicator()}
    </div>
  );
};

export default Carousel;
