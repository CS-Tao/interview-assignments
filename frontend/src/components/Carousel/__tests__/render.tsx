import { render } from "@testing-library/react";
import React from "react";
import Carousel from "..";

test("render", () => {
  const { baseElement } = render(
    <Carousel
      items={[
        {
          title: "xPhone",
          subTitle: ["Lots to love. Less to spend.", "Starting at $399."],
          image: "iphone",
          backColor: "#111111",
          foreColor: "white",
        },
        {
          title: "Tablet",
          subTitle: "Just the right amount of everything.",
          image: "tablet",
          backColor: "#fafafa",
          foreColor: "black",
        },
        {
          title: ["Buy a Tablet or xPhone for college.", "Get arPods"],
          image: "airpods",
          backColor: "#f2f2f3",
          foreColor: "black",
        },
      ]}
    />
  );
  expect(baseElement).toMatchSnapshot();
});
