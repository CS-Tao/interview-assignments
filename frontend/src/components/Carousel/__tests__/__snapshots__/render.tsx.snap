// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render 1`] = `
<body>
  <div>
    <div
      class="carousel-wrapper"
      style="width: 100%; height: 100%;"
    >
      <div
        class="carousel-item"
        style="color: white; background-color: rgb(17, 17, 17); transition: left 500ms linear; left: 0%;"
      >
        <div
          class="title-wrapper"
        >
          <div>
            xPhone
          </div>
        </div>
        <div
          class="subtitle-wrapper"
        >
          <div>
            Lots to love. Less to spend.
          </div>
          <div>
            Starting at $399.
          </div>
        </div>
        <div
          class="image-wrapper"
        >
          <img
            alt=""
            src="iphone"
          />
        </div>
      </div>
      <div
        class="carousel-item"
        style="color: black; background-color: rgb(250, 250, 250); transition: left 500ms linear; left: 100%;"
      >
        <div
          class="title-wrapper"
        >
          <div>
            Tablet
          </div>
        </div>
        <div
          class="subtitle-wrapper"
        >
          <div>
            Just the right amount of everything.
          </div>
        </div>
        <div
          class="image-wrapper"
        >
          <img
            alt=""
            src="tablet"
          />
        </div>
      </div>
      <div
        class="carousel-item"
        style="color: black; background-color: rgb(242, 242, 243); transition: left 500ms linear; left: 200%;"
      >
        <div
          class="title-wrapper"
        >
          <div>
            Buy a Tablet or xPhone for college.
          </div>
          <div>
            Get arPods
          </div>
        </div>
        <div
          class="subtitle-wrapper"
        />
        <div
          class="image-wrapper"
        >
          <img
            alt=""
            src="airpods"
          />
        </div>
      </div>
      <div
        class="carousel-indicator"
      >
        <div
          class="carousel-indicator-item"
        >
          <div
            class="carousel-indicator-item-handler"
            style="width: 100%; transition: width 2500ms linear;"
          />
        </div>
        <div
          class="carousel-indicator-item"
        >
          <div
            class="carousel-indicator-item-handler"
            style="width: 0%;"
          />
        </div>
        <div
          class="carousel-indicator-item"
        >
          <div
            class="carousel-indicator-item-handler"
            style="width: 0%;"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
